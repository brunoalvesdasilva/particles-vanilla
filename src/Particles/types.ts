export interface IParticles {}

export type IPosition = {
  x: number
  y: number
}
