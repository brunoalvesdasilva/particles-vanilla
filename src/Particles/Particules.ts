import { IParticles, IPosition } from './types'

class Particles implements IParticles {
  private x: number
  private y: number
  private radius: number
  private color: string
  private speed: number
  private moveX: number = 0
  private moveY: number = 0

  constructor (
    x: number,
    y: number,
    radius: number,
    speed: number,
    color: string = '#000000'
  ) {
    this.x = x
    this.y = y
    this.radius = radius
    this.color = color
    this.speed = speed

    this.calculateMoviment()
  }

  getRadius () {
    return this.radius
  }

  calculateMoviment () {
    const isPositiveMoveX = this.moveX >= 0
    const isPositiveMoveY = this.moveY >= 0

    this.moveX = isPositiveMoveX ? this.speed : this.speed * -1
    this.moveY = isPositiveMoveY ? this.speed : this.speed * -1
  }

  getNextPosition (): IPosition {
    return { x: this.x + this.moveX, y: this.y + this.moveY }
  }

  setPosition ({ x, y }: IPosition) {
    this.x = x
    this.y = y
  }

  getPosition (): IPosition {
    return { x: this.x, y: this.y }
  }

  flipX () {
    this.moveX = this.moveX * -1
  }

  flipY () {
    this.moveY = this.moveY * -1
  }

  move () {
    const { x, y } = this.getNextPosition()

    this.x = x
    this.y = y
  }

  build (cxt: CanvasRenderingContext2D) {
    cxt.beginPath()
    cxt.arc(this.x, this.y, this.radius, 0, Math.PI * 2)
    cxt.fillStyle = this.color
    cxt.fill()
    cxt.closePath()
  }
}

export default Particles
