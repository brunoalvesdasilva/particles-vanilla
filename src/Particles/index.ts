export { default as Particle } from './Particules'
export * from './types'
