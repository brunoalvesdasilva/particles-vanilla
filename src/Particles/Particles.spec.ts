import { describe, expect, it } from 'vitest'
import Particles from './Particules'

describe('Particles', () => {
  it('should return instanceof Particles class', () => {
    const particles = new Particles(0, 0, 0, 0)

    expect(particles).instanceOf(Particles)
  })

  it('should move particle to right and down with speed is 1', () => {
    const expected = { x: 2, y: 2 }
    const particles = new Particles(0, 0, 0, 1)

    particles.move()

    expect(particles.getNextPosition()).toMatchObject(expected)
  })

  it('should move particle to left and top with speed is -1', () => {
    const expected = { x: 0, y: 0 }
    const particles = new Particles(2, 2, 0, -1)

    particles.move()

    expect(particles.getNextPosition()).toMatchObject(expected)
  })

  it('should return new position when method setPosition is called', () => {
    const expectedBeforeCalled = { x: 0, y: 0 }
    const expectedAfterCalled = { x: 10, y: 10 }
    const particles = new Particles(0, 0, 0, -1)

    expect(particles.getPosition()).toMatchObject(expectedBeforeCalled)

    particles.setPosition({ x: 10, y: 10 })

    expect(particles.getPosition()).toMatchObject(expectedAfterCalled)
  })

  it('should return radius 8 when particle is created with radius 8', () => {
    const expected = 8
    const particles = new Particles(0, 0, 8, 0)

    expect(particles.getRadius()).toBe(expected)
  })

  it('should flip position y when mehotd flipY is called', () => {
    const expected = { x: 1, y: 3 }
    const particles = new Particles(2, 2, 0, -1)

    particles.flipY()

    expect(particles.getNextPosition()).toMatchObject(expected)
  })

  it('should flip position x when mehotd flipX is called', () => {
    const expected = { x: 3, y: 1 }
    const particles = new Particles(2, 2, 0, -1)

    particles.flipX()

    expect(particles.getNextPosition()).toMatchObject(expected)
  })

  it('should build canvas context with information particles', () => {
    const fakeCanvas = document.createElement('canvas')
    const fakeContext = fakeCanvas.getContext('2d')!

    const particles = new Particles(2, 2, 0, -1)

    particles.build(fakeContext)

    // @ts-ignore
    const path = fakeContext.__getPath()
    expect(path).toMatchSnapshot()
  })
})
