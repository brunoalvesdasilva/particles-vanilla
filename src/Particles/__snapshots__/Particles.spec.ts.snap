// Vitest Snapshot v1

exports[`Particles > should build canvas context with information particles 1`] = `
[
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  {
    "props": {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 0,
      "startAngle": 0,
      "x": 2,
      "y": 2,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "closePath",
  },
]
`;
